### PR TITLE
Hugging Face Inference API Integration

### PR DESCRIPTION
**Pull request by MinuraPunchihewa**
Hey @ZoranPandovski, @paxcema,

I have implemented an initial version of the integration for the Hugging Face Inference API. For now, it supports the following tasks,

- text-classification (sentiment analysis)
- fill-mask
- summarization

I will work on supporting additional tasks as well.

Given below is an example of how we can use it to execute a summarization workload,

```
CREATE ML_ENGINE hf_inference_api_engine
FROM huggingface_inference_api
USING
  api_key = 'hf_XXXXXXXXXXXXXXXXXXXXX';

CREATE MODEL mindsdb.hf_summarizer
PREDICT summary
USING
    task = 'summarization',
    engine = 'hf_inference_api_engine',
    api_key = 'hf_XXXXXXXXXXXXXXXXXXXXX',
    inputs = {
      "column": "text"
    };

SELECT text, summary
FROM mindsdb.hf_summarizer
WHERE text = 'The tower is 324 metres (1,063 ft) tall, about the same height as an 81-storey building, and the tallest structure in Paris. Its base is square, measuring 125 metres (410 ft) on each side. During its construction, the Eiffel Tower surpassed the Washington Monument to become the tallest man-made structure in the world, a title it held for 41 years until the Chrysler Building in New York City was finished in 1930. It was the first structure to reach a height of 300 metres. Due to the addition of a broadcasting aerial at the top of the tower in 1957, it is now taller than the Chrysler Building by 5.2 metres (17 ft). Excluding transmitters, the Eiffel Tower is the second tallest free-standing structure in France after the Millau Viaduct.';
```
I'd love to hear your thoughts on this.

### CODE REVIEW

File: mindsdb/integrations/handlers/huggingface_inference_api_handler/__init__.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+from mindsdb.integrations.handlers.autosklearn_handler.__about__ import __version__ as version, __description__ as description
+try:
+    from .huggingface_inference_api_handler import HuggingFaceInferenceAPIHandler as Handler
+    import_error = None
+except Exception as e:
+    Handler = None
+    import_error = e
+
+title = 'Hugging Face Inference API'
+name = 'huggingface_inference_api'
```
Comment (paxcema) [inline mindsdb/integrations/handlers/huggingface_inference_api_handler/__init__.py:12]: Perhaps shorten to `huggingface_api`?

File: mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    Integration with the Hugging Face Inference API.
+    """
+
+    name = 'huggingface_inference_api'
+
+    def create(self, target: str, df: Optional[pd.DataFrame] = None, args: Optional[Dict] = None) -> None:
+        if 'using' not in args:
+            raise Exception("Hugging Face Inference engine requires a USING clause! Refer to its documentation for more details.")
+
+        config = ConfigParser(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'config.yaml'))
```
Comment (paxcema) [inline mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py:24]: Shouldn't the user be able to pass these params via `USING`? In particular, the model name seems quite important.

File: mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+    def predict(self, df: Optional[pd.DataFrame] = None, args: Optional[Dict] = None) -> None:
+        args = self.model_storage.json_get('args')
+        config_args = self.model_storage.json_get('config_args')
+
+        inputs = self._parse_inputs(df, args['using']['inputs'], args['using']['task'])
+
+        response = self._query(
+            f"{config_args['BASE_URL']}/{config_args['TASK_MODEL_MAP'][args['using']['task']]}",
+            args['using']['api_key'],
```
Comment (paxcema) [inline mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py:38]: Much like in the OpenAI handler, would be nice to have multiple cases where this key can come from: USING args, but also env vars or at engine creation time.

**Review by paxcema — CHANGES_REQUESTED** [review]
General thoughts: looks amazing, this will be super useful. 

I do wonder if we require an entire new handler for it, but having both is definitely better than having just one without inference API support.

Left a few nitpicks, the only other important thing that would be cool is adding a few tests here and there, if possible.

Otherwise, this looks good!

**Comment by paxcema** [comment]
Also, forgot to ask: can you please fill in [this form](https://docs.google.com/document/d/1yOpRkfjhtL6KmueC0dIvn6rZ-PD3b9XdIa7Ywnyp6Jk/edit) and upload a `README.md` with the answers?

File: mindsdb/integrations/handlers/huggingface_inference_api_handler/__init__.py (hunk 1 above)
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/huggingface_inference_api_handler/__init__.py:12]: Yes, that sounds better. Done!

File: mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py (hunk 2 above)
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py:24]: Yes, the users can pass a model of their choice now. If it is not provided, the recommend model will be used.

File: mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py (hunk 3 above)
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/huggingface_inference_api_handler/huggingface_inference_api_handler.py:38]: Done!

**Comment by MinuraPunchihewa** [comment]
I will add some tests and fill the form soon, Patricio.

**Comment by MinuraPunchihewa** [comment]
I thought I'd drop some of the commands I ran to test out the handler here,

```
CREATE ML_ENGINE hf_api_engine
FROM huggingface_api
USING
  api_key = 'hf_...';

# text classification
CREATE MODEL mindsdb.hf_sentiment_classifier
PREDICT sentiment
USING
  task = 'text-classification',
  column = 'text',
  engine = 'hf_api_engine',
  api_key = 'hf_...'

SELECT text, sentiment
FROM mindsdb.hf_sentiment_classifier_22
WHERE text = 'I like you. I love you.';

# fill mask
CREATE MODEL mindsdb.hf_fill_mask
PREDICT sequence
USING
  task = 'fill-mask',
  column = 'text',
  engine = 'hf_api_engine',
  api_key = 'hf_...'

SELECT text, sequence
FROM mindsdb.hf_fill_mask
WHERE text = 'The answer to the universe is [MASK].';

# summarization
CREATE MODEL mindsdb.hf_summarizer
PREDICT summary
USING
  task = 'summarization',
  engine = 'hf_api_engine',
  api_key = 'hf_...',
  column = 'text'

SELECT text, summary
FROM mindsdb.hf_summarizer
WHERE text = 'The tower is 324 metres (1,063 ft) tall, about the same height as an 81-storey building, and the tallest structure in Paris. Its base is square, measuring 125 metres (410 ft) on each side. During its construction, the Eiffel Tower surpassed the Washington Monument to become the tallest man-made structure in the world, a title it held for 41 years until the Chrysler Building in New York City was finished in 1930. It was the first structure to reach a height of 300 metres. Due to the addition of a broadcasting aerial at the top of the tower in 1957, it is now taller than the Chrysler Building by 5.2 metres (17 ft). Excluding transmitters, the Eiffel Tower is the second tallest free-standing structure in France after the Millau Viaduct.';

# image classification
CREATE MODEL mindsdb.hf_image_classifier
PREDICT classification
USING
  task = 'image-classification',
  engine = 'hf_api_engine',
  api_key = 'hf_...',
  column = 'image'

SELECT image, classification
FROM mindsdb.hf_image_classifier
WHERE image = 'https://raw.githubusercontent.com/MinuraPunchihewa/hugging-py-face/main/tests/resources/dogs.jpeg';

# speech recognition
CREATE MODEL mindsdb.hf_speech_recognition
PREDICT transcript
USING
  task = 'speech-recognition',
  engine = 'hf_api_engine',
  api_key = 'hf_...',
  column = 'audio'

SELECT audio, transcript
FROM mindsdb.hf_speech_recognition
WHERE audio = 'https://raw.githubusercontent.com/MinuraPunchihewa/hugging-py-face/main/tests/resources/amused.wav';

# audio-classification
CREATE MODEL mindsdb.hf_audio_classifier
PREDICT classification
USING
  task = 'audio-classification',
  engine = 'hf_api_engine,
  api_key = 'hf_...',
  column = 'audio'

SELECT audio, classification
FROM mindsdb.hf_audio_classifier
WHERE audio = 'https://raw.githubusercontent.com/MinuraPunchihewa/hugging-py-face/main/tests/resources/amused.wav';
```

**Comment by MinuraPunchihewa** [comment]
Hey @ZoranPandovski and @paxcema,
Apologies, it looks like I have forgotten to push the commits making up the new implementation of the handler. They are available now. I have also added a `README.md` based on the document given above.

File: mindsdb/integrations/handlers/huggingface_api_handler/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
+# Briefly describe what ML framework does this handler integrate to MindsDB, and how?
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/huggingface_api_handler/README.md:1]: This is for internal review. @paxcema Should this be added in the handler dir?

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]

**Comment by MinuraPunchihewa** [comment]
Hey @paxcema,
As discussed, I have also added support for the [question answering, sentence similarity and zero shot classification tasks.

**Comment by ZoranPandovski** [comment]
@MinuraPunchihewa can you make sure this is up to date with staging, so we can solve the failing tests?

**Pull request merged**